Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -540,28 +540,37 @@ extension HTTPHeaders: RandomAccessCollection {
     }
 }
 
-extension Character {
+extension UTF8.CodeUnit {
     var isASCIIWhitespace: Bool {
-        return self == " " || self == "\t" || self == "\r" || self == "\n" || self == "\r\n"
+        switch self {
+        case UInt8(ascii: " "),
+             UInt8(ascii: "\t"),
+             UInt8(ascii: "\r"),
+             UInt8(ascii: "\n"):
+          return true
+
+        default:
+          return false
+        }
     }
 }
 
 extension String {
     func trimASCIIWhitespace() -> Substring {
-        return self.dropFirst(0).trimWhitespace()
+        return Substring(self).trimWhitespace()
     }
 }
 
-private extension Substring {
-    func trimWhitespace() -> Substring {
-        var me = self
-        while me.first?.isASCIIWhitespace == .some(true) {
-            me = me.dropFirst()
-        }
-        while me.last?.isASCIIWhitespace == .some(true) {
-            me = me.dropLast()
+extension Substring {
+    fileprivate func trimWhitespace() -> Substring {
+        guard let firstNonWhitespace = self.utf8.firstIndex(where: { !$0.isASCIIWhitespace }) else {
+          // The whole substring is ASCII whitespace.
+          return Substring()
         }
-        return me
+
+        // There must be at least one non-ascii whitespace character, so banging here is safe.
+        let lastNonWhitespace = self.utf8.lastIndex(where: { !$0.isASCIIWhitespace })!
```

Comment:
I don't believe this force-unwrap is safe: we can in principle decode an empty string, and while that's not really valid in our specific use-case we should probably have some defense in depth against that bug making its way into lower areas of the code, or with other areas.

Reply:
I don't see how this is unsafe given `firstNonWhitespace` is known to not be `nil` at this point: we're checking for the same predicate but from the other direction?

Reply:
Oh apologies, you're right. I missed that we were duplicating the check.